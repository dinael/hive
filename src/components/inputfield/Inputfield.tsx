import {FC, useState, ChangeEvent, FocusEvent } from 'react'
import styles from './Inputfield.module.scss'
import Icon from '../icon/Icon'
import ValidationMessage from '../validationmessage/ValidationMessage'
import { Label as LabelText } from '../label/Label'

export type InputfieldProps = {
  label: string
  labelHidden?: boolean
  id?: string
  name?: string
  type?: 'text' | 'email' | 'password' | 'search' | 'number' | 'tel' | 'url' | 'date' | 'datetime-local' | 'time' | 'datetime' | 'mouth' | 'week' | 'year'
  placeholder?: string
  value?: string | number
  onBlur?: (event: FocusEvent<HTMLInputElement>) => void
  onFocus?: (event: FocusEvent<HTMLInputElement>) => void
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void
  error?: boolean
  errorText?: string
  readOnly?: boolean
  disabled?: boolean
  required?: boolean
  isRequiredText?: string
  maxLength?: number
  className?: string
}

export const Inputfield: FC<InputfieldProps> = ({
  label = '',
  labelHidden = false,
  id,
  name,
  type = 'text',
  placeholder = '',
  value,
  onBlur,
  onFocus,
  onChange,
  error = false,
  errorText,
  readOnly,
  disabled,
  required = false,
  isRequiredText,
  maxLength,
  className,
  ...props
}: InputfieldProps) => {

  const [isFocused, setIsFocused] = useState(false)
  const [hasValue, setValue] = useState<string | number>('')
  const [hasError, setError] = useState('')

  const handleFocus = (event: FocusEvent<HTMLInputElement>) => {
    setIsFocused(true)
    onFocus && onFocus(event)
  }

  const handleBlur = (event: FocusEvent<HTMLInputElement>) => {
    setIsFocused(false)
    onBlur && onBlur(event)
  }

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setValue(event.target.value)
    onChange && onChange(event)
  }

  // TODO: ADD ICON
  return (
    <label
      htmlFor={id}
      className={`
        ${styles['inputfiled-component']}
        ${error? styles['has-error'] : ''}
      `}
      {...props}>
      <LabelText
        onlyText
        className={`${styles['inputfiled-label']}`}
        text={label}
        required={required}
        isRequiredText={`${isRequiredText}`}
        labelHidden={labelHidden}/>
      <input
        className={error && !isFocused ? 'error' : ''}
        id={id}
        name={name}
        type={type ? type : 'text'}
        placeholder={placeholder}
        value={hasValue}
        required={required}
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        maxLength={maxLength}
        readOnly={readOnly}
        disabled={disabled}
        aria-describedby={`validation-message-${id}`}
      />
      {error && !isFocused ?
        <ValidationMessage
          className={styles['input-field-validation-message']}
          message={`${errorText}`}
          kind='error'
          role="alert"
          aria-live="assertive"
          aria-atomic="true"
          id={`validation-message-${id}`}/>
      : null}
    </label>
  )
}

export default Inputfield
