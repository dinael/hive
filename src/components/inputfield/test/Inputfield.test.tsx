import { describe, test, expect, vi} from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'
import { Inputfield, InputfieldProps } from '../Inputfield'

describe('Inputfield', () => {
  const defaultProps: InputfieldProps = {
    label: 'Label',
    labelHidden: false,
    id: 'id-test',
    name: 'name',
    type: 'text',
    placeholder: 'Placeholder',
    value: 'value-test',
    onChange: vi.fn(),
    onBlur: vi.fn(),
    onFocus: vi.fn(),
    error: false,
    errorText: 'error-test',
    required: false,
    disabled: false,
    readOnly: false,
    isRequiredText: 'isRequiredText-test',
    maxLength: undefined,
    className: 'className-test',
  }

  test('renders snapshot', () => {
    const { container } = render(<Inputfield {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

})
