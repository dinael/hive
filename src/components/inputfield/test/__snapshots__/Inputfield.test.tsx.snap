// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Inputfield > renders snapshot 1`] = `
<div>
  <label
    class="
        inputfiled-component
        
      "
    for="id-test"
  >
    <span
      class="
      label-text-component
      
      
      
      inputfiled-label
      "
    >
      Label:
    </span>
    <input
      aria-describedby="validation-message-id-test"
      class=""
      id="id-test"
      name="name"
      placeholder="Placeholder"
      type="text"
      value=""
    />
  </label>
</div>
`;
